src/cpu_feature/arm_linux/arm_linux.rs

```diff
@@ -0,0 +1,307 @@
+use std::collections::HashSet;
+use std::string::{String, ToString};
+use c::ulong;
+#[cfg(all(any(target_arch = "arm", target_arch = "aarch64"), target_os="linux"))]
+use self::auxv::AuxValError;
+use self::auxv::AuxVals;
+use self::cpuinfo::CpuInfo;
+#[cfg(all(any(target_arch = "arm", target_arch = "aarch64"), target_os="linux"))]
+use self::cpuinfo::parse_cpuinfo;
+
+// Bits exposed in HWCAP and HWCAP2
+// See /usr/include/asm/hwcap.h on an ARM installation for the source of
+// these values.
+const ARM_HWCAP2_AES: ulong = 1 << 0;
+const ARM_HWCAP2_PMULL: ulong = 1 << 1;
+const ARM_HWCAP2_SHA1: ulong = 1 << 2;
+const ARM_HWCAP2_SHA2: ulong = 1 << 3;
+
+const ARM_HWCAP_NEON: ulong = 1 << 12;
+
+// Constants used in GFp_armcap_P
+// from include/openssl/arm_arch.h
+#[cfg(any(test, all(any(target_arch = "arm", target_arch = "aarch64"), target_os="linux")))]
+const ARMV7_NEON: u32 = 1 << 0;
+// not a typo; there is no constant for 1 << 1
+const ARMV8_AES: u32 = 1 << 2;
+const ARMV8_SHA1: u32 = 1 << 3;
+const ARMV8_SHA256: u32 = 1 << 4;
+const ARMV8_PMULL: u32 = 1 << 5;
+
+extern "C" {
+    #[cfg(all(any(target_arch = "arm", target_arch = "aarch64"), target_os="linux"))]
+    #[allow(non_upper_case_globals)]
+    pub static mut GFp_armcap_P: u32;
+}
+
+#[cfg(all(any(target_arch = "arm", target_arch = "aarch64"), target_os="linux"))]
+#[allow(non_snake_case)]
+extern "C" fn GFp_cpuid_setup() {
+    if let Ok(c) = parse_cpuinfo() {
+        // TODO handle failures to read from procfs auxv
+        if let Ok(auxvals) = auxv::search_auxv(&Path::from("/proc/self/auxv"),
+                   &[auxv::AT_HWCAP, auxv::AT_HWCAP2]) {
+            let armcap = arm_cpuid_setup(&c, &auxvals);
+            unsafe {
+                GFp_armcap_P |= armcap;
+            }
+        }
+    }
+}
+
+/// returns the GFp_armcap_P bitstring
+#[cfg(any(test, all(any(target_arch = "arm", target_arch = "aarch64"), target_os="linux")))]
+#[allow(dead_code)] // TODO
+fn arm_cpuid_setup(cpuinfo: &CpuInfo, procfs_auxv: &AuxVals) -> u32 {
+    let mut hwcap: ulong = 0;
+
+    // |getauxval| is not available on Android until API level 20. If it is
+    // unavailable, read from /proc/self/auxv as a fallback. This is unreadable
+    // on some versions of Android, so further fall back to /proc/cpuinfo.
+    // See
+    // https://android.googlesource.com/platform/ndk/+/882ac8f3392858991a0e1af33b4b7387ec856bd2
+    // and b/13679666 (Google-internal) for details. */
+
+    // TODO link to getauxval weakly and use that if available. In particular,
+    // this is needed by the assumption in hwcap_from_cpuinfo that future architectures
+    // will have a working getauxval.
```

Comment:
For `getauxval()`, I recommend just defining a wrapper around it in C code that returns an error code if the function isn't available.